gcp/apis/compute/v1alpha1/network_types.go

```diff
@@ -0,0 +1,297 @@
+// In the future, this can be generated automatically.
+type GCPNetworkSpec struct {
+	// IPv4Range: Deprecated in favor of subnet mode networks. The range of
+	// internal addresses that are legal on this network. This range is a
+	// CIDR specification, for example: 192.168.0.0/16. Provided by the
+	// client when the network is created.
+	IPv4Range string `json:"IPv4Range,omitempty"`
+
+	// AutoCreateSubnetworks: When set to true, the VPC network is created
+	// in "auto" mode. When set to false, the VPC network is created in
+	// "custom" mode. When set to nil, the VPC network is created in "legacy"
+	// mode which will be deprecated by GCP soon.
+	//
+	// An auto mode VPC network starts with one subnet per region. Each
+	// subnet has a predetermined range as described in Auto mode VPC
+	// network IP ranges.
+	AutoCreateSubnetworks *bool `json:"autoCreateSubnetworks,omitempty"`
+
+	// Description: An optional description of this resource. Provide this
+	// field when you create the resource.
+	Description string `json:"description,omitempty"`
+
+	// Name: Name of the resource. Provided by the client when the resource
+	// is created. The name must be 1-63 characters long, and comply with
+	// RFC1035. Specifically, the name must be 1-63 characters long and
+	// match the regular expression `[a-z]([-a-z0-9]*[a-z0-9])?. The first
+	// character must be a lowercase letter, and all following characters
+	// (except for the last character) must be a dash, lowercase letter, or
+	// digit. The last character must be a lowercase letter or digit.
+	Name string `json:"name,omitempty"`
+
+	// RoutingConfig: The network-level routing configuration for this
+	// network. Used by Cloud Router to determine what type of network-wide
+	// routing behavior to enforce.
+	RoutingConfig *GCPNetworkRoutingConfig `json:"routingConfig,omitempty"`
+}
+
+// IsSameWith compares the fields of GCPNetworkSpec and
+// GCPNetworkStatus to report whether there is a difference. Its cyclomatic
+// complexity is related to how many fields exist, so, not much of an indicator.
+// nolint:gocyclo
+func (in GCPNetworkSpec) IsSameWith(n GCPNetworkStatus) bool {
+	if (in.RoutingConfig != nil && n.RoutingConfig == nil) ||
+		(in.RoutingConfig == nil && n.RoutingConfig != nil) {
+		return false
+	}
+	if in.RoutingConfig != nil && n.RoutingConfig != nil && in.RoutingConfig.RoutingMode != n.RoutingConfig.RoutingMode {
+		return false
+	}
+	if (in.AutoCreateSubnetworks == nil && n.AutoCreateSubnetworks) ||
+		(in.AutoCreateSubnetworks != nil && *in.AutoCreateSubnetworks != n.AutoCreateSubnetworks) {
+		return false
+	}
+	if in.Description != n.Description ||
+		in.Name != n.Name ||
+		in.IPv4Range != n.IPv4Range {
+		return false
+	}
+	return true
+}
+
+// GCPNetworkStatus is the complete mirror of googlecompute.Network but
+// with deepcopy functions. In the future, this can be generated automatically.
+type GCPNetworkStatus struct {
```

Comment:
> If so, it's convention in the Kubernetes API to update the writable fields in spec, rather than to mirror them in the status.

Currently, generic managed reconciler doesn't natively support updating anything other than `status` subresource. All updates are made via `r.client.Status().Update(ctx, managed)`. One can go ahead and make the k8s client available in `Observe` and then update the resource but that's a complete hack, we need to implement a nice way to do this https://github.com/crossplaneio/crossplane-runtime/issues/10

Reply:
I'm concerned that these two types introduce a pattern in which we mirror the entire object, rather than just the "status related" fields in the status object. Perhaps we do want to introduce this pattern going forward, but if we shipped these two types as is for 0.3 they'd look quite different from all of our other managed resources.

Would it be feasible to include on the fields relevant to the resource status under `.status` for 0.3 and figure out how we want this to work more broadly for 0.4?

Reply:
IMO, all fields are indeed status related, including the ones that we have in `spec` as well. For example, `RoutingConfig` exists in both types and if you change it in the `spec` I think it's good that you can make sure the change is reflected on the provider by looking at the status. Maybe `Kind` that is same for all Network CRs or immutable fields like `Name` are exempt from this. But if we think of the `status` as the place that all other resource will use to get information from that CR, including immutable fields seems useful.

Overall, I'd like us to make the least number of decisions possible when implementing a managed kind so that the higher level components can make use of any field they want. In its claim, we can omit some fields or add new ones that are bound to more than one field via a custom logic. But for the managed I think limiting `spec` to only what you can change and having `status` as everything you can ever see gives more flexibility to higher level components as well as different resource usage scenarios without going back and change things on the managed resource. That does make some fields, which don't seem very useful to have like `Kind`, included but I think that's a verbosity that is OK to have on that level for overall consistency.

Reply:
Nevertheless, I'll remove the fields that don't look status-related to have consistent user experience for CRs as you pointed out.

Reply:
We discussed this in standup this morning. Removing all spec fields from the status would require some not-insignificant changes to the logic of this PR, so we're going to merge this as is and consider removing them later. In the meantime we need to ensure we raise an issue to decide whether we want to adopt this pattern.